### PR TITLE
Add stale-contact-point-timeout setting to ClusterBootstrap

### PR DESCRIPTION
**Pull request by Arkatufus**
## Changes

Add a new HOCON settings called stale-contact-point-timeout to control how long a bootstrap contact point be retained until it is considered as stale and needs to be refreshed by the HTTP probes.

### CODE REVIEW

File: src/management/Akka.Management/Cluster/Bootstrap/ClusterBootstrapSettings.cs [hunk 1]
```diff
@@ -152,27 +152,39 @@ public static ContactPointSettings Create(Config config)
...
                 var fallbackPort = string.IsNullOrWhiteSpace(fallback) || fallback == "<fallback-port>"
                     ? (int?) null : int.Parse(fallback);
+                var probeFailureTimeout = contactPointConfig.GetTimeSpan("probing-failure-timeout", null, false);
+                var staleEntryTimeoutStr = contactPointConfig.GetString("stale-contact-point-timeout");
+                var staleEntryTimeout = string.IsNullOrWhiteSpace(staleEntryTimeoutStr)
+                                        || staleEntryTimeoutStr is "off"
+                                        || staleEntryTimeoutStr is "false"
+                                        || staleEntryTimeoutStr is "no"
+                        ? probeFailureTimeout
+                        : contactPointConfig.GetTimeSpan("stale-contact-point-timeout");
```
Comment (Arkatufus) [inline src/management/Akka.Management/Cluster/Bootstrap/ClusterBootstrapSettings.cs:162]: backward compatibility mode, if this setting is not set (off, false, no, or empty), use the old behavior of using the probe failure timeout value

File: src/management/Akka.Management/Cluster/Bootstrap/Internal/BootstrapCoordinator.cs [hunk 2]
```diff
@@ -159,6 +160,7 @@ public BootstrapCoordinator(ServiceDiscovery discovery, IJoinDecider joinDecider
             _discovery = discovery;
             _joinDecider = joinDecider;
             _settings = settings;
+            _staleContactPointTimeout = _settings.ContactPoint.ProbeInterval + _settings.ContactPoint.StaleContactPointTimeout;
```
Comment (Arkatufus) [inline src/management/Akka.Management/Cluster/Bootstrap/Internal/BootstrapCoordinator.cs:163]: This is where the new setting being used, actual value is a combination of the setting value and the probe interval value

File: src/management/Akka.Management/Cluster/Bootstrap/Internal/BootstrapCoordinator.cs [hunk 3]
```diff
@@ -484,8 +486,7 @@ private void Decide()
             
             // filter out old observations, in case the probing failures are not triggered
             bool IsObsolete(SeedNodesObservation obs)
-                => (currentTime - obs.ObservedAt).TotalMilliseconds >
-                   _settings.ContactPoint.ProbingFailureTimeout.TotalMilliseconds;
+                => (currentTime - obs.ObservedAt).TotalMilliseconds > _staleContactPointTimeout.TotalMilliseconds;
```
Comment (Arkatufus) [inline src/management/Akka.Management/Cluster/Bootstrap/Internal/BootstrapCoordinator.cs:489]: Actual stale/obsolete calculation for contact points, the stale timeout value is used here

File: src/management/Akka.Management/Cluster/Bootstrap/Internal/HttpContactPointBootstrap.cs [hunk 4]
```diff
@@ -94,6 +95,8 @@ public HttpContactPointBootstrap(ClusterBootstrapSettings settings, ResolvedTarg
             _probeInterval = settings.ContactPoint.ProbeInterval;
             _probeRequest = ClusterBootstrapRequests.BootstrapSeedNodes(baseUri);
             _cancellationTokenSource = new CancellationTokenSource();
+
+            _probingFailureTimeout = _probeInterval + _settings.ContactPoint.ProbingFailureTimeout;
```
Comment (Arkatufus) [inline src/management/Akka.Management/Cluster/Bootstrap/Internal/HttpContactPointBootstrap.cs:99]: Probe timeout value is modified to take consideration the actual probing interval time span

File: src/management/Akka.Management/Resources/reference.conf [hunk 5]
```diff
@@ -174,6 +174,13 @@ akka.management {
       # If some discovered seed node will keep failing to connect for specified period of time,
       # it will initiate rediscovery again instead of keep trying.
       probing-failure-timeout = 3s
+      
+      # Set the time for bootstrap coordinator to consider that previously discovered contact point
+      # were considered to be stale and needs to be re-discovered by the HTTP probes.
+      #
+      # The legacy value for this is equals to the value of `probing-failure-timeout`. This is the value
+      # that will be used if `stale-contact-point-timeout` is set to `off`, `false`, `no`, or empty.
+      stale-contact-point-timeout = off
```
Comment (Arkatufus) [inline src/management/Akka.Management/Resources/reference.conf]: New HOCON setting

**Review by Arkatufus — COMMENTED** [review]
self review

**Comment by Arkatufus** [comment]
Cluster formation average time for a 22 node cluster, using Akka.Management.Cluster.Bootstrap, varying the probe interval and the new stale contact point timeout setting.

Cluster formation average time, control group:
image

Cluster formation average time, 5 second probe interval:
image

Cluster formation average time, 10 seconds probe interval:
image

File: src/management/Akka.Management/Resources/reference.conf (hunk 5 above)
Comment (Aaronontheweb) [reply src/management/Akka.Management/Resources/reference.conf]: The default value is `probing-failure-timeout` PLUS the `probe-interval`, correct?

**Review by Aaronontheweb — COMMENTED** [review]
Left a question for you on the new configuration setting

File: src/management/Akka.Management/Resources/reference.conf (hunk 5 above)
Comment (Arkatufus) [reply src/management/Akka.Management/Resources/reference.conf]: yes, that is correct

**Review by Aaronontheweb — APPROVED** [review]
LGTM

**Pull request merged**